### DISCUSSION TITLE
I need to use HA-ModBusProxy

### DISCUSSION THREAD

**Discussion by TMoersch** [wlcrs/huawei_solar#266] [Setting up the integration]
Hi,
how can I change the inverters IP in the configuration if needed??
Because I want to run huwaei_solar an evcc parallel, I need to use modbud-proxy to serialize the modbus requests.
The addon is already installed and I want huawei_solar to communicate over the modbud-proxy?

Thomas

**Comment by scrichab** [comment]
deinstall and reinstall huawei_solar
i´ve done that a few times testing ModBus Proxy

**Reply by TMoersch** [reply]
maybe I'am stupid but I dont understand how the communication flow works whith the information I got.....
I know how squid is working ( I know its not the same function here, buts in my head when I think about proxy commincation flow).
You normaly configure your proxy to listen to your local LAN IP adresse an set the listen port. 
In this case any apllication that can connect the proxy over IP:port is able to use the proxy, but still need to be informed to do so!!
The application (huawei / evcc) need the configuration which end-device is to call (OK), but they do not have configuration statements which external proxy is to use ??? 
In huawei_solar I can't define which proxy to use??
How is this working? like export_proxy:...   as an environmental variable?
evcc seem to have it's own buildin modbus_proxy. If I follow the configuration info I got, evcc does start something that is listening on port 502. But this seems to be unreachable for huawei_solar, now that I deletet my Huawei installation I cant set them up again because ther is no communcation to the inverter???? 
In the documentation of modbus-proxy is defined to set the IP to the inverters IP??
I think the listening interface should be something like localhost or local interface IP.
If several independend application are sending the request to proxy IP:port, the proxy can do his job.

I believe I'am totally wrong in my head an can't find any documantation that explains how this should be setup properly.

I realy would love to have the evcc nad huawei_solar parallel in use, communicatiing vie modbus-proxy, this would realy maximise my benefit of having solar and ev.

```
s6-rc: info: service legacy-services: starting
s6-rc: info: service legacy-services successfully started
Preparing to run modbus-proxy
Upstream: 192.168.2.223:502
Listen: 502
Generated Config
devices:
  - modbus:
      url: 192.168.2.223:502 # device url (mandatory)
      timeout: 10 # communication timeout (s) (optional, default: 10)
      connection_time: 0.1 # delay after connection (s) (optional, default: 0)
    listen:
      bind: 0:502 # listening address (mandatory)
2023-01-21 11:19:23,523     INFO modbus-proxy: Starting...
2023-01-21 11:19:23,536     INFO modbus-proxy.ModBus(192.168.2.223:502): Ready to accept requests on 0:502
```
Here is what AddOn HA-modbus-proxy ist telling me in his log, for me it looks fine.

**Reply by malgates** [reply]
I am also stuck here. Did not find a way to fix it

`Logger: huawei_solar.huawei_solar
Quelle: /usr/local/lib/python3.12/site-packages/huawei_solar/huawei_solar.py:348
Erstmals aufgetreten: 10:36:07 (34 Vorkommnisse)
Zuletzt protokolliert: 13:24:31
Modbus client is not connected to the inverter.

NoneType: None
`

**Reply by malgates** [reply]
'Dieser Fehler wurde von einer benutzerdefinierten Integration verursacht

Logger: backoff
Quelle: custom_components/huawei_solar/config_flow.py:115
Integration: Huawei Solar (Dokumentation, Probleme)
Erstmals aufgetreten: 10:38:48 (7 Vorkommnisse)
Zuletzt protokolliert: 13:24:31

Giving up _do_read(...) after 6 tries (huawei_solar.exceptions.ConnectionInterruptedException: Modbus client is not connected to the inverter.)`

**Reply by malgates** [reply]
'Logger: huawei_solar.huawei_solar
Quelle: /usr/local/lib/python3.12/site-packages/huawei_solar/huawei_solar.py:184
Erstmals aufgetreten: 10:38:48 (7 Vorkommnisse)
Zuletzt protokolliert: 13:24:31

Aborting client creation due to error.
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/backoff/_async.py", line 151, in retry
    ret = await target(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/backoff/_async.py", line 151, in retry
    ret = await target(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/huawei_solar/huawei_solar.py", line 349, in _do_read
    raise ConnectionInterruptedException(message)
huawei_solar.exceptions.ConnectionInterruptedException: Modbus client is not connected to the inverter.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/huawei_solar/huawei_solar.py", line 178, in create
    await huawei_solar._initialize()
  File "/usr/local/lib/python3.12/site-packages/huawei_solar/huawei_solar.py", line 107, in _initialize
    self.time_zone = (await self.get(rn.TIME_ZONE)).value
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/huawei_solar/huawei_solar.py", line 241, in get
    return (await self.get_multiple([name], slave))[0]
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/huawei_solar/huawei_solar.py", line 276, in get_multiple
    response = await self._read_registers(registers[0].register, total_length, slave)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/huawei_solar/huawei_solar.py", line 398, in _read_registers
    return await _do_read()
           ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/backoff/_async.py", line 159, in retry
    await _call_handlers(on_giveup, **details, exception=e)
  File "/usr/local/lib/python3.12/site-packages/backoff/_async.py", line 37, in _call_handlers
    await handler(details)
  File "/usr/local/lib/python3.12/site-packages/backoff/_async.py", line 16, in f
    return coro_or_func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/huawei_solar/huawei_solar.py", line 327, in backoff_giveup
    raise ReadException(f"Failed to read register {register} after {details['tries']} tries")
huawei_solar.exceptions.ReadException: Failed to read register 43006 after 6 tries
'
